README.md
# Polkastarter Lottery

`polkastarter-lottery` is a gem that generates a list of winners that can participate on a Polkastarter Pool.
This is based on a set of criteria that aim to turn access to a Pool the most democratic and fair as possible, as well as transparent and fail-proof.

This gem is used by Polkastarter to generate lotteries.

Check [this blog post](https://blog.polkastarter.com/polkastarter-whitelists-just-got-a-whole-lot-better/) to understand the full mechanic for this lottery system.

# Installation

`gem install polkastarter-lottery`


# Basic usage

```ruby
balances = { '0x111' => 3000, '0x222' => 1_000, '0x333' => 30_000 }

service = LotteryService.new balances: balances,
                             max_winners: 1_000,
                             seed: 1234567890 # optional
service.run
service.winners
```

# Run specs

```
bundle install
rspec
```

# Default settings

##### Price per ticket

`250 POLS`

##### Weights for each holder

```
*  < 250  POLS -> 0.00
*    250+ POLS -> 1.00
*  1 000+ POLS -> 1.10
*  3 000+ POLS -> 1.15
* 10 000+ POLS -> 1.20
* 30 000+ POLS -> 1.25
```

##### Top 10 holders

Top 10 holders are always winners. They can always particiapte at any pool and also bypass the cool down period.

##### Reproduce previous lottery results

1. Download 'applications.csv' from polkastarter.com, containing the structure per row: application_id,pols_power
2. Place 'applications.csv' file in the root of project
3. Build and install the gem with `gem build && gem install polkastarter-lottery`
4. Run `./verify <max_winners> <application_id> <seed>`


CHANGELOG.md
# v1.3.1: Bugfix on verify script

* Bug fix: the "verify" script was still using the legacy "application_id" instead of "identifier"

# 1.3.0: 26.Apr.2022: Lottery transparency

* Completely refactor lottery for simplicity and tranparency
* Allow to replicate the lottery with a random seed number (generated in a trustworthy way and stored on-chain) on the user side for full transparency
* Simplify tests and better tests

# 1.2.2: 18.Jan.2022: Disable never-winning ratio by default

* Disable never-winning ratio by default, as we're not using it anymore for a long time

# 1.2.1: 4.Jan.2022: Improvement: Guard against invalid balances

* Add guard against invalid balances

# 1.2.0: 17 Nov 2021: Add support for NFTs

* Feature: Add support for NFTs

# 1.1.2: 22.Sep.2021: Minor bugfix

* Bugfix: Do not count as top holder if the participants list is smallers than 10

# 1.1.1: 23.Jun.2021: Minor bugfix

* Bugfix: Fix a minor issue when there are no winners, a [nil] was returned.

# 1.1.0: 18.Jun.2021: Expose input params for better flexibility

* Expose `#top_n_holders` and `#privileged_never_winning_ratio` on input API

# 1.0.1: 20.May.2021: Expose more output data for stats

* Expose `#top_holders` and `#privileged_participants` (shuffled 10% of users that never won before)

# 1.0.0: 6.May.2021: Initial Release

* Initial release


spec/spec_helper.rb
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration

Dir[File.expand_path('lib/**/*.rb'), __FILE__].each { |file| require file }

require 'initializers/enumerable'
require 'initializers/hash'
require 'shared/theoretical_probabilites_example'
require 'matchers/be_around'

RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/initializers/hash.rb
class Hash
  # Returns a new hash with +self+ and +other_hash+ merged recursively.
  #
  #   h1 = { a: true, b: { c: [1, 2, 3] } }
  #   h2 = { a: false, b: { x: [3, 4, 5] } }
  #
  #   h1.deep_merge(h2) # => { a: false, b: { c: [1, 2, 3], x: [3, 4, 5] } }
  #
  # Like with Hash#merge in the standard library, a block can be provided
  # to merge values:
  #
  #   h1 = { a: 100, b: 200, c: { c1: 100 } }
  #   h2 = { b: 250, c: { c1: 200 } }
  #   h1.deep_merge(h2) { |key, this_val, other_val| this_val + other_val }
  #   # => { a: 100, b: 450, c: { c1: 300 } }
  def deep_merge(other_hash, &block)
    dup.deep_merge!(other_hash, &block)
  end

  # Same as +deep_merge+, but modifies +self+.
  def deep_merge!(other_hash, &block)
    merge!(other_hash) do |key, this_val, other_val|
      if this_val.is_a?(Hash) && other_val.is_a?(Hash)
        this_val.deep_merge(other_val, &block)
      elsif block_given?
        block.call(key, this_val, other_val)
      else
        other_val
      end
    end
  end
end


spec/matchers/be_around.rb
RSpec.configure do |config|
  RSpec::Matchers.define :be_around do |expected_value, error|
    match do |actual_value|
      actual_value >= expected_value - error &&
        actual_value <= expected_value + error
    end

    failure_message do |actual_value|
      "expected that #{actual_value} would be around [#{expected_value - error}, #{expected_value + error}] with a margin error of #{error}, but it failed by #{(actual_value - expected_value).abs}"
    end
  end
end


lib/services/lottery_service.rb
#require_relative '../models/participant'
#require File.dirname(__FILE__) + '/../models/participant'
require 'models/participant'
require 'pry'

class LotteryService
  attr_reader :balances           # e.g: { '0x71C7656EC7ab88b098defB751B7401B5f6d8976F' => 3000 }
  attr_reader :recent_winners     # e.g: ['0x71C7656EC7ab88b098defB751B7401B5f6d8976F']
  attr_reader :past_winners       # e.g: ['0x71C7656EC7ab88b098defB751B7401B5f6d8976F']
  attr_reader :blacklist          # e.g: ['0x71C7656EC7ab88b098defB751B7401B5f6d8976F']
  attr_reader :nft_rare_holders   # e.g: ['0x71C7656EC7ab88b098defB751B7401B5f6d8976F'] # rare NFT
  attr_reader :nft_common_holders # e.g: ['0x71C7656EC7ab88b098defB751B7401B5f6d8976F'] # common NFT

  attr_reader :all_participants # all candidates
  attr_reader :eligibles        # only eligible ones
  attr_reader :participants     # only the list of wallets that will be actually shuffled (i.e. excluding top holders and nft rare holders)
  attr_reader :winners          # only winners
  attr_reader :max_winners
  attr_reader :top_n_holders
  attr_reader :privileged_never_winning_ratio
  attr_reader :top_holders
  attr_reader :privileged_participants

  DEFAULT_MAX_WINNERS = 1_000.freeze
  DEFAULT_TOP_N_HOLDERS = 10.freeze
  DEFAULT_PRIVILEGED_NEVER_WINNING_RATIO = 0.0.freeze

  def initialize(balances:,
                 max_winners: DEFAULT_MAX_WINNERS,
                 top_n_holders: DEFAULT_TOP_N_HOLDERS,
                 privileged_never_winning_ratio: DEFAULT_PRIVILEGED_NEVER_WINNING_RATIO,
                 recent_winners: [],
                 past_winners: [],
                 blacklist: [],
                 nft_rare_holders: [],
                 nft_common_holders: [])
    @balances = balances
    @max_winners = max_winners
    @top_n_holders = top_n_holders
    @privileged_never_winning_ratio = privileged_never_winning_ratio
    @recent_winners = recent_winners
    @past_winners = past_winners.map &:downcase
    @blacklist = blacklist
    @nft_rare_holders = nft_rare_holders
    @nft_common_holders = nft_common_holders
  end

  def run
    @all_participants = build_participants.sort # sort desc by balance
    @eligibles        = all_participants.select(&:eligible?)
    @top_holders      = @eligibles.first top_n_holders
    @participants     = @eligibles.reject do |participant|
      top_holder?(participant) || participant.nft_rare_holder # top holders and nft1 holders are always excluded from shuffling because they will always enter
    end.sort # also order them by balance

    @privileged_participants = shuffled_privileged_participants

    @winners = calculate_winners.compact
  end

  # Only used for statistics and debugging
  def stats_by_tier
    Participant::BALANCE_WEIGHTS.keys.inject({}) do |tiers, tier|
      tier_participants = participants.select { |participant| participant.tier == tier }.size
      tier_winners      = winners.select      { |winner| winner.tier == tier }.size

      tiers[tier] = {
        winners:      tier_winners,
        participants: tier_participants,
        percentage:   tier_winners.to_f / tier_participants
      }
      tiers
    end
  end

  private

  def calculate_winners
    winners = []

    winners += top_holders
    winners += nft_rare_participants
    winners += privileged_participants
    winners += shuffled_eligible_participants

    winners.uniq.first(max_winners)
  end

  def top_holder?(participant)
    top_holders.include? participant.address
  end

  def shuffled_privileged_participants
    sample_size = max_winners * privileged_never_winning_ratio
    never_winning_participants.sample sample_size
  end

  def weighted_random_sample(participants)
    participants.max_by do |participant|
      rand ** (1.0 / participant.tickets)
    end
  end

  def shuffled_eligible_participants
    return participants if participants.count <= max_winners

    to_shuffle = participants
    winners = []
    while winners.count < participants.count do
      winners += max_winners.times.map { weighted_random_sample(to_shuffle) }
      winners.uniq!

      to_shuffle = to_shuffle - winners
    end

    winners
  end

  def never_winning_participants
    participants.select { |participant| !past_winners.include? participant.address }
  end

  def nft_rare_participants
    eligibles.select { |participant| participant.nft_rare_holder }
  end

  def build_participants
    nft_rare_balances = nft_rare_holders.map { |addr| [addr, nil] }.to_h
    all_balances = balances.merge(nft_rare_balances)

    all_balances.map do |address, balance|
      next if blacklist.include? address
      Participant.new address:           address,
                      balance:           balance,
                      recent_winner:     recent_winners.include?(address),
                      nft_rare_holder:   nft_rare_holders.include?(address),
                      nft_common_holder: nft_common_holders.include?(address)
    end.compact
  end
end


lib/models/participant.rb
require 'date'

class Participant
  attr_reader :identifier
  attr_reader :balance

  def initialize(identifier:, balance:, xorshift: nil)
    @identifier = identifier
    @balance    = (balance && balance >= 0) ? balance : 0 # guarding against invalid balances
    @xorshift   = xorshift || XorshiftService.new
  end

  TICKET_PRICE = 250.freeze # e.g: 100 means 1 ticket = 100 POLS
  BALANCE_WEIGHTS = {       # e.g: { 1000 => 1.1 } means 1000 POLS weigths 1.1
    0      => 0.00,
    250    => 1.00,
    1_000  => 1.10,
    3_000  => 1.15,
    10_000 => 1.20,
    30_000 => 1.25
  }.freeze

  def dice
    @dice ||= @xorshift.next_float
  end

  def probability
    dice ** (1.0 / tickets) # reservoir sampling
  end

  def tickets
    (balance / TICKET_PRICE).to_i * weight
  end

  def weight
    @weight ||= BALANCE_WEIGHTS[tier]
  end

  def tier
    BALANCE_WEIGHTS.keys.select { |min_balance| min_balance <= balance }.last
  end

  def eligible?
    tickets > 0
  end

  def <=>(other)
    other.probability <=> self.probability
  end
end
